src/monte_carlo.c

```diff
@@ -0,0 +1,185 @@
+#include "monte_carlo.h"
+#include "linear_congruential_random_generator.h"
+#include <math.h>
+#include <stdlib.h>
+#include <string.h>
+#include <time.h>
+
+mc_real calc_UCB(Node* node) {
+  if (node->nVisits == 0) {
+    return UCB_MAX;
+  }
+  return node->score / node->nVisits +
+         sqrt(2 * log10(node->parent->nVisits) / node->nVisits);
+}
+
+Node* find_max_UCB(Node* children, unsigned nChildren) {
+  Node* max_ucb_node = children;
+  for (int i = 0; i < nChildren; i++) {
+    if (calc_UCB(children + i) > calc_UCB(max_ucb_node)) {
+      max_ucb_node = children + i;
+    }
+  }
+  return max_ucb_node;
+}
+
+Node* select_node(Node* node) {
+  Node* curr_node = node;
+  while (curr_node->children) {
+    curr_node = find_max_UCB(curr_node->children, curr_node->nChildren);
+    if (calc_UCB(curr_node) == UCB_MAX) {
+      return curr_node;
+    }
+  }
+  return curr_node;
+}
+
+void expand_leaf(Node* node, int player, Game game) {
+  int nPossibleActions = game.getNumPossibleActions(node->state);
+  Action possibleActions[nPossibleActions];
+  game.getPossibleActions(node->state, possibleActions);
+
+  int nValidActions = 0;
+  for (int i = 0; i < nPossibleActions; i++) {
+    if (!game.isValidAction(&(node->state), possibleActions + i, player)) {
+      nValidActions++;
+    }
+  }
+
+  // Deep copy of parent
+  node->children = (struct Node*)malloc(nValidActions * sizeof(struct Node));
+
+  nValidActions = 0;
+  for (int i = 0; i < nPossibleActions; ++i) {
+    if (!game.isValidAction(&(node->state), possibleActions + i, player)) {
+      ((node->children) + nValidActions)->nVisits = 0;
+      ((node->children) + nValidActions)->score = 0;
+
+      // Deep copy of board
+      Board board = game.playAction(node->state, possibleActions + i);
+      ((node->children) + nValidActions)->state.values =
+          malloc(game.getBoardSize() * sizeof(int));
+      memcpy(((node->children) + nValidActions)->state.values, board.values,
+             game.getBoardSize() * sizeof(int));
+      ((node->children) + nValidActions)->state.nPlayers = board.nPlayers;
```

Comment:
Oof, the same computation is executed multiple times. I think we should use a tmp variable to help the compiler know that it's safe to store this address in a regsiter. (node->children + nValidActions)